### PR TITLE
[pylint] R1711: useless-return

### PR DESCRIPTION
**Pull request by tomecki**
This PR implements https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/useless-return.html , including autofixing.

Impl change from original approach: the check locates the actual useless return statements, in contrast to pylint's implemnentation which flags the _functions_ which have useless return statements.

### CODE REVIEW

**Comment by tomecki** [comment]
Root ticket for pylint coverage: https://github.com/charliermarsh/ruff/issues/970

File: crates/ruff/src/rules/pylint/rules/useless_return.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                Ok(fix) => {
+                    diagnostic.amend(fix);
+                }
+                Err(e) => {
+                    error!("Failed to delete `return` statement: {}", e);
+                }
+            };
+        }
+        checker.diagnostics.push(diagnostic);
+    }
```
Comment (youknowone) [inline crates/ruff/src/rules/pylint/rules/useless_return.rs:85]: returning here might fit better with another one in line 57-59
```suggestion
    if !is_bare_return_or_none {
        return;
    }
    let mut diagnostic = Diagnostic::new(UselessReturn, Range::from_located(stmt));
    if !checker.patch(diagnostic.kind.rule()) {
        match delete_stmt(
            stmt,
            None,
            &[],
            checker.locator,
            checker.indexer,
            checker.stylist,
        ) {
            Ok(fix) => {
                diagnostic.amend(fix);
            }
            Err(e) => {
                error!("Failed to delete `return` statement: {}", e);
            }
        };
    }
    checker.diagnostics.push(diagnostic);
```

**Comment by charliermarsh** [comment]
I think the one issue with this rule is that it may conflict with some of the `flake8-return` rules, e.g. those that _require_ an explicit `return None` at the end of the function, like `RET502` and `RET503`. I'm hesitant to introduce rules that are in direct conflict. An alternative would be to make the `flake8-return` rules parameterizable such that they can enforce this behavior if desired, in lieu of enforcing the opposite.

**Comment by r3m0t** [comment]
Looks like in pylint the rule ignores functions which can return values.  https://github.com/PyCQA/pylint/blob/1eef2273aee4c5a2e287f7470d6da3a3ae7d0760/pylint/checkers/refactoring/refactoring_checker.py#L2022

Which makes it the same as/very similar to R501 - https://pypi.org/project/flake8-return/

**Comment by charliermarsh** [comment]
Oh, in that case, it may actually be different than any existing checks? Since none of those tell you to remove a `return` entirely, only that the return should be implicit or explicit in various cases. (`R501` only tells you to change `return None` to `return` if all returns are `None`, not to drop a `return` / `return None` at the end of a method if it's the only `return`.)

**Comment by charliermarsh** [comment]
But we need to adjust this such that it only warns if this is the _only_ return in the function.

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**